pkg/ddl/reorg_util.go

```diff
@@ -0,0 +1,216 @@
+		}
+		m.SetMaxWriteSpeed(int(vardef.DDLReorgMaxWriteSpeed.Load()))
+	}
+
+	if setDistTaskParam {
+		m.IsDistReorg = vardef.EnableDistTask.Load()
+		m.IsFastReorg = vardef.EnableFastReorg.Load()
+		m.TargetScope = dxfhandle.GetTargetScope()
+		if kerneltype.IsNextGen() {
+			m.MaxNodeCount = scheduler.CalcMaxNodeCountByTableSize(ctx, tableSizeInBytes)
+		} else {
+			//nolint:forbidigo
+			if sv, ok := sctx.GetSessionVars().GetSystemVar(vardef.TiDBMaxDistTaskNodes); ok {
+				m.MaxNodeCount = variable.TidbOptInt(sv, 0)
+				if m.MaxNodeCount == -1 { // -1 means calculate automatically
+					m.MaxNodeCount = scheduler.CalcMaxNodeCountByStoresNum(ctx, sctx.GetStore())
+				}
+			}
+		}
+
+		if hasSysDB(job) {
+			if m.IsDistReorg {
+				logutil.DDLLogger().Info("cannot use distributed task execution on system DB",
+					zap.Stringer("job", job))
+			}
+			m.IsDistReorg = false
+			m.IsFastReorg = false
+			failpoint.Inject("reorgMetaRecordFastReorgDisabled", func(_ failpoint.Value) {
+				LastReorgMetaFastReorgDisabled = true
+			})
+		}
+		if m.IsDistReorg && !m.IsFastReorg {
+			return dbterror.ErrUnsupportedDistTask
+		}
+	}
+	job.ReorgMeta = m
+	logutil.DDLLogger().Info("initialize reorg meta",
+		zap.String("jobSchema", job.SchemaName),
+		zap.String("jobTable", job.TableName),
+		zap.Stringer("jobType", job.Type),
+		zap.Bool("enableDistTask", m.IsDistReorg),
+		zap.Bool("enableFastReorg", m.IsFastReorg),
+		zap.String("targetScope", m.TargetScope),
+		zap.Int("maxNodeCount", m.MaxNodeCount),
+		zap.Int("concurrency", m.GetConcurrency()),
+		zap.Int("batchSize", m.GetBatchSize()),
+	)
+	return nil
+}
+
+func getTableSizeByID(ctx context.Context, store kv.Storage, tblID int64) int64 {
+	helperStore, ok := store.(helper.Storage)
+	if !ok {
+		logutil.DDLLogger().Error("store does not implement helper.Storage interface",
+			zap.String("storeType", fmt.Sprintf("%T", store)))
+		return 0
+	}
+	h := helper.NewHelper(helperStore)
+	pdCli, err := h.TryGetPDHTTPClient()
+	if err != nil {
+		logutil.DDLLogger().Error("failed to get PD HTTP client for calculating table size",
+			zap.Int64("tableID", tblID),
+			zap.Error(err))
+		return 0
+	}
+	totalSize, err := estimateTableSizeByID(ctx, pdCli, tblID)
+	if err != nil {
+		logutil.DDLLogger().Error("failed to estimate table size for calculating concurrency",
+			zap.Int64("tableID", tblID),
+			zap.Error(err))
+	}
+	if totalSize == 0 {
+		regionStats, err := h.GetPDRegionStats(ctx, tblID, false)
+		if err != nil {
+			logutil.DDLLogger().Error("failed to get region stats for calculating concurrency",
+				zap.Int64("tableID", tblID),
+				zap.Error(err))
+			return 0
+		}
+		return regionStats.StorageSize * units.MiB
+	}
+	return totalSize
+}
+
+func estimateTableSizeByID(ctx context.Context, pdCli pdhttp.Client, pid int64) (int64, error) {
+	sk, ek := tablecodec.GetTableHandleKeyRange(pid)
+	sRegion, err := pdCli.GetRegionByKey(ctx, codec.EncodeBytes(nil, sk))
```

Comment:
why we cannot use the previous `sk, ek` directly, and use the one get from PD region?

Reply:
Not sure if there are specific reasons, this logic is copied from `getRegionsInfoForSingleTable`. Maybe I can ask the author later.